Please ignore the pull request: it's just an experiment for the #7 fix

vikulin opened the pull request:
1. fixed issue for Yggdrasil().Stop() crash: see #7

neilalexander commented on src/mobile/mobile.go:
```diff
@@ -4,12 +4,12 @@ import (
 	hjson "github.com/hjson/hjson-go"
 	"github.com/mitchellh/mapstructure"
-	"github.com/yggdrasil-network/yggdrasil-extras/src/dummy"
+	"github.com/vikulin/yggdrasil-extras/src/dummy"
```
Need to revert this import change.

vikulin replied:
Got it. Reverted.

vikulin commented:
@neilalexander , thank you for the quick reply. I have never considered the change will go to live code. I have reverted the import code line which I changed while the experiment.

vikulin commented:
One note: the os.Exit(0) call invokes app shutdown. Could we re-name the method from **Stop()** to Exit(rc) and add **Stop()** without os.Exit which could continue Yggdrasil instance running but without traffic, multicasting and admin part. This would require to add corresponding **Resume()** method: it unfreeze the Yggdrasil() instance activity

neilalexander commented:
So I actually really don’t like that it calls `os.Exit()` at all, because the idea is that Yggdrasil can be embedded as a library in another application. If we call `os.Exit()` then we kill the host application when we maybe don’t want to.

We should be able to clean up the Yggdrasil node without killing the entire process, which is the purpose of the `Stop()` function. If there are genuine problems where we close the same channel twice then I would consider that to be a bug, but this is not really a fix.

vikulin commented:
@neilalexander , agreed. The crash reproduces I guess because `Stop()` functions are run twice:

https://github.com/yggdrasil-network/yggdrasil-extras/pull/8/files#diff-07935d1f9a6eab5fa08bfa6a808040beL114
https://github.com/yggdrasil-network/yggdrasil-extras/pull/8/files#diff-07935d1f9a6eab5fa08bfa6a808040beL115

 Shell I test the code part without `os.Exit()` or you have an idea how to fix it quickly?

```
func (m *Yggdrasil) Stop() {
	m.multicast.Stop()
	m.admin.Stop()
	m.core.Stop()
	m.core.Stop()
}
```

vikulin commented:
@neilalexander , One more variant: would the Yggdrasil be locked by the infinite cycle here?

https://github.com/yggdrasil-network/yggdrasil-go/blob/a9cfa5bc0dd41032efbb11fe4a843ff90a6ee73f/src/yggdrasil/tcp.go#L217

```
	for {
		sock, err := l.Listener.Accept()
		if err != nil {
			t.link.core.log.Errorln("Failed to accept connection:", err)
			return
		}
		t.waitgroup.Add(1)
		go t.handler(sock, true, nil, l.upgrade)
	}
```

vikulin commented on src/mobile/mobile.go:
```diff
@@ -110,12 +109,14 @@ func (m *Yggdrasil) StartJSON(configjson []byte) (conduit *dummy.ConduitEndpoint
+	logger := log.New(m.log, "", 0)
+	logger.EnableLevel("info")
+	logger.Infof("Stop the mobile Yggdrasil instance %s", "")
+	m.admin.Stop()
+	m.multicast.Stop()
+	m.dummy.Stop()
 	m.core.Stop()
```
finally, the code worked for me

neilalexander commented on src/mobile/mobile.go:
```diff
@@ -110,12 +110,14 @@ func (m *Yggdrasil) StartJSON(configjson []byte) (conduit *dummy.ConduitEndpoint
+	logger.Infof("Stop the mobile Yggdrasil instance %s", "")
+	//m.admin.Stop()
+	m.multicast.Stop()
+	//m.dummy.Stop()
```
Why wouldn't we want the dummy adapter to be stopped?

vikulin replied:
Could you please continue the optimization in other commit? I would like to focus on Android and Kotlin development.

neilalexander commented on src/mobile/mobile.go:
```diff
@@ -110,12 +110,14 @@ func (m *Yggdrasil) StartJSON(configjson []byte) (conduit *dummy.ConduitEndpoint
 // Stop the mobile Yggdrasil instance
-func (m *Yggdrasil) Stop() error {
+func (m *Yggdrasil) Stop() {
+	logger := log.New(m.log, "", 0)
```
Would be better to factor this out into the `Yggdrasil` struct so we reuse a single logger, rather than creating more of them.

neilalexander commented on src/dummy/dummy.go:
```diff
@@ -152,8 +152,8 @@ func (dummy *DummyAdapter) _start() error {
 	return nil
 }
 
-// Start the setup process for the Dummy adapter. If successful, starts the
```
What happened to this comment?
